openedx/frontend-app-course-authoring
# Fix  video upload failures

## Description

This is a fix for a bug where video uploads are endlessly "In Progress" or "Uploading". I was able to reproduce this problem locally by navigating away from the videos page while a video upload was happening. That resulted in 1. The frontend sending a request to the edx backend to mark a video upload as in-progress and get an S3 upload URL, 2. the frontend starting a S3 upload, 3. the S3 upload being interrupted, producing a non-2xx response from S3 and explaining the "Incomplete Body" response we saw in our logs, and 4. the logic that tells the edx backend to mark the upload as failed never being run because the page is closed already and the code execution is interrupted.

This fix attempts to achieve 3 things.
1. users should be made aware that a frontend upload is in progress and it's not safe to navigate away from the page
2. users should not see any "Uploading" video statuses from existing videos that never finished their uploads and thus failed. They should know they failed so they can retry the upload.
3. when users navigate away from a page and interrupt an upload, this should not cause the backend status of the video to remain "Uploading" but it should become "Failed".

This is how these problems are fixed:
1. While any video upload is in progress, a very visible banner is displayed at the top of the page. See screenshot. The UX needs review.

Note on the screenshot: The black toast with "Adding 0 videos" only appears for 3 seconds, while the banner stays there while any video upload is in progress. Also, is the "0" in the toast a bug? Depends on whether this also happens on stage/prod with real video uploads. Anyway I would consider this out of scope.

![Screenshot 2024-04-12 at 2 28 30 PM](https://github.com/openedx/frontend-app-course-authoring/assets/19345795/9cae5048-60f9-45ae-9673-91e3bcef3913)


2. Just in the frontend, every "Uploading" status label is changed to "Failed". This is because there's in reality no background uploads - there's only an active upload while the user is on the page. When the user opens the video page for the time, it's impossible for there be an upload in progress already, so that label is misleading. My understanding is also that when you started an upload, the card only loads anyway when the upload finished or failed, or when the user refreshes the page, so there should never be an "Uploading" label that's accurate.

3. before page unload, a request is sent to the backend to change every in-progress video upload's status to 'Failed'.

## Questions to reviewers

- should Toast that appears at the same time as the banner, saying e.g. "1 Upload started" be removed?
- for simplicity's sake, I changed the status display for the tag for all failed statuses to "Failed". So 'Failed', 'Partial Failure', 'Uploading' all are displayed as 'Failed'. Should I change this so 'Partial Failure' is displayed as such?

## Supporting information

[2U JIRA Task](https://2u-internal.atlassian.net/browse/TNL-11532)

## Testing instructions

Testing needs a couple of temporary code changes:
- enable the waffle flag `contentstore.mock_video_uploads`
- For the change in the labels from 'Uploading' to 'Failed', you need to have a course with some Videos in the 'Uploading' state. This is not easy to get, so I would add this line ONLY TEMPORARILY to your local `edx-platform//cms/djangoapps/contentstore/video_storage_handlers.py` inside the _get_index_videos inside the `_get_values` function:
`values['status'] = 'Uploading'`. Add it right underneath the TODO comment you see in this code and do not forget to remove it after testing:

```
    def _get_values(video, course):
        """
        Get data for predefined video attributes.
        """
        values = {}
        for attr in attrs:
            # bunch of existing code, whatever
        # TODO: remove next line! For testing only - the next line is the one you should insert here
        values['status'] = 'Uploading'
        return values

    videos, pagination_context = _get_videos(course, pagination_conf)
    return [_get_values(video, course) for video in videos], pagination_context
```

- For testing the upload alert popping up, uncomment these lines in `src/files-and-videos/videos-page/data/thunks.js` - I will remove them before merging:
```
      // sleep 10 seconds
      // await new Promise((resolve) => setTimeout(resolve, 10000));
```
- then, just click the button to upload a video. You should see the alert popping up for 10 seconds, then disappear.
- When navigating away from the page, for each video upload in progress, the backend should receive a request to set the video's upload status to failed:
- Still keep the sleep lines from above uncommented
- start uploading a video and then immediately refresh the page or change the url
- Go to the network tab. You should see the request sent to the backend that explicitly sets the video status to failed (should be the one that returns the 500, assuming you mocked out the video uploads using the waffle flag). If the request is there and it has the correct data that tells the backend to set the upload status to failed, the logic is working.
